operators/katib-controller/reactive/katib_controller.py

```diff
@@ -0,0 +1,394 @@
+
+
+@when("charm.started")
+def charm_ready():
+    layer.status.active("")
+
+
+@when("layer.docker-resource.oci-image.changed")
+def update_image():
+    clear_flag("charm.started")
+
+
+def gen_certs(namespace, service_name):
+    if Path("/run/cert.pem").exists():
+        hookenv.log("Found existing cert.pem, not generating new cert.")
+        return
+
+    Path("/run/ssl.conf").write_text(
+        f"""[ req ]
+default_bits = 2048
+prompt = no
+default_md = sha256
+req_extensions = req_ext
+distinguished_name = dn
+[ dn ]
+C = GB
+ST = Canonical
+L = Canonical
+O = Canonical
+OU = Canonical
+CN = 127.0.0.1
+[ req_ext ]
+subjectAltName = @alt_names
+[ alt_names ]
+DNS.1 = {service_name}
+DNS.2 = {service_name}.{namespace}
+DNS.3 = {service_name}.{namespace}.svc
+DNS.4 = {service_name}.{namespace}.svc.cluster
+DNS.5 = {service_name}.{namespace}.svc.cluster.local
+IP.1 = 127.0.0.1
+[ v3_ext ]
+authorityKeyIdentifier=keyid,issuer:always
+basicConstraints=CA:FALSE
+keyUsage=keyEncipherment,dataEncipherment,digitalSignature
+extendedKeyUsage=serverAuth,clientAuth
+subjectAltName=@alt_names"""
+    )
+
+    check_call(["openssl", "genrsa", "-out", "/run/ca.key", "2048"])
+    check_call(["openssl", "genrsa", "-out", "/run/server.key", "2048"])
+    check_call(
+        [
+            "openssl",
+            "req",
+            "-x509",
+            "-new",
+            "-sha256",
+            "-nodes",
+            "-days",
+            "3650",
+            "-key",
+            "/run/ca.key",
+            "-subj",
+            "/CN=127.0.0.1",
+            "-out",
+            "/run/ca.crt",
+        ]
+    )
+    check_call(
+        [
+            "openssl",
+            "req",
+            "-new",
+            "-sha256",
+            "-key",
+            "/run/server.key",
+            "-out",
+            "/run/server.csr",
+            "-config",
+            "/run/ssl.conf",
+        ]
+    )
+    check_call(
+        [
+            "openssl",
+            "x509",
+            "-req",
+            "-sha256",
+            "-in",
+            "/run/server.csr",
+            "-CA",
+            "/run/ca.crt",
+            "-CAkey",
+            "/run/ca.key",
+            "-CAcreateserial",
+            "-out",
+            "/run/cert.pem",
+            "-days",
+            "365",
+            "-extensions",
+            "v3_ext",
+            "-extfile",
+            "/run/ssl.conf",
+        ]
+    )
+
+
+@when("layer.docker-resource.oci-image.available")
+@when_not("charm.started")
+def start_charm():
+    if not hookenv.is_leader():
+        hookenv.log("This unit is not a leader.")
+        return False
+
+    layer.status.maintenance("configuring container")
+
+    image_info = layer.docker_resource.get_info("oci-image")
+    namespace = os.environ["JUJU_MODEL_NAME"]
+    config = dict(hookenv.config())
+
+    gen_certs(namespace, hookenv.service_name())
+    ca_bundle = b64encode(Path("/run/cert.pem").read_bytes()).decode("utf-8")
```

Comment:
`cert-manager` isn't necessary for these operators, since they're able to do things like creating certificates, as above.

I'm not opposed to `katib-controller` being capable of dynamic webhook creation, as long as there's a way to disable it. I ran into issues with Juju and `katib-controller` fighting over editing the dynamic resources.

I could open this up as a PR if you'd like, to allow disabling of dynamic webhook creation: https://github.com/knkski/katib/commit/833c1bb3ffb1acbebe7e3d302b941662b941f8a9

Reply:
@knkski If we disable automatically webhook creation in the controller and prepare manifests, will it work for Juju?
As I can see, the problem is how we generate cert for the webhook server.
I created an issue, we can continue discussion there: https://github.com/kubeflow/katib/issues/1405.